File: java-kata/src/test/java/com/leetcode/ThreeSumTest.java
package com.leetcode;

import org.junit.jupiter.api.Test;

import java.util.*;

import static org.assertj.core.api.Assertions.assertThat;
import static test.ArrayUtil.arrayOf;

class ThreeSumTest {

    private final ThreeSum it = new ThreeSum();

    @Test
    void test0() {
        var result = it.threeSum(arrayOf(-1, 0, 1, 2, -1, -4));
        assertThat(result)
                .containsExactlyInAnyOrder(List.of(-1, -1, 2), List.of(-1, 0, 1));
    }

    @Test
    void test1() {
        var result = it.threeSum(arrayOf(0, 1, 1));
        assertThat(result).isEmpty();
    }

    @Test
    void test2() {
        var result = it.threeSum(arrayOf(0, 0, 0));
        assertThat(result)
                .isEqualTo(List.of(List.of(0, 0, 0)));
    }

    @Test
    void test3() {
        var result = it.threeSum(arrayOf(-4,-2,-2,-2,0,1,2,2,2,3,3,4,4,6,6));
        assertThat(result)
                .containsExactlyInAnyOrder(
                        List.of(-4,-2,6),
                        List.of(-4,0,4),
                        List.of(-4,1,3),
                        List.of(-4,2,2),
                        List.of(-2,-2,4),
                        List.of(-2,0,2)
                );
    }
}

File: kotlin-kata/src/test/kotlin/com/leetcode/ThreeSumTest.kt
package com.leetcode

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test

internal class ThreeSumTest {

    private val target = ThreeSum()

    @Test
    fun test0() {
        val result = target.threeSum(intArrayOf(-1, 0, 1, 2, -1, -4))
        assertThat(result)
            .containsExactlyInAnyOrder(listOf(-1, -1, 2), listOf(-1, 0, 1))
    }

    @Test
    fun test1() {
        val result = target.threeSum(intArrayOf(0, 1, 1))
        assertThat(result)
            .isEmpty()
    }

    @Test
    fun test2() {
        val result = target.threeSum(intArrayOf(0, 0, 0))
        assertThat(result)
            .containsExactlyInAnyOrder(listOf(0, 0, 0))
    }

    @Test
    fun test3() {
        val result = target.threeSum(intArrayOf(-4, -2, -2, -2, 0, 1, 2, 2, 2, 3, 3, 4, 4, 6, 6))
        assertThat(result)
            .containsExactlyInAnyOrder(
                listOf<Int>(-4, -2, 6),
                listOf<Int>(-4, 0, 4),
                listOf<Int>(-4, 1, 3),
                listOf<Int>(-4, 2, 2),
                listOf<Int>(-2, -2, 4),
                listOf<Int>(-2, 0, 2)
            )
    }
}